scalafix/rules/src/main/scala/migrate/MigrationRule.scala

```diff
@@ -0,0 +1,153 @@
+package migrate
+
+import metaconfig.Configured
+import scalafix.patch.Patch
+import scalafix.util.TokenOps
+import scalafix.v1._
+import utils.ScalaExtensions._
+
+import scala.meta._
+import scala.meta.contrib.Trivia
+import scala.meta.internal.pc.ScalafixGlobal
+import scala.meta.tokens.Token
+import scala.tools.nsc.reporters.StoreReporter
+import scala.util.{Failure, Success}
+
+class MigrationRule(global: ScalafixGlobal) extends SemanticRule("MigrationRule") {
+  override def description: String = "infer types and show synthetics"
+
+  def this() = this(null)
+
+  override def withConfiguration(config: Configuration): Configured[Rule] = {
+    if (config.scalacClasspath.isEmpty) {
+      Configured.error(s"config.scalacClasspath should not be empty")
+    } else {
+      val global = CompilerService.newGlobal(config.scalacClasspath, config.scalacOptions)
+      global match {
+        case Success(settings) => Configured.ok(new MigrationRule(new ScalafixGlobal(settings, new StoreReporter, Map())))
+        case Failure(exception) => Configured.error(exception.getMessage)
+      }
+    }
+  }
+
+  override def fix(implicit doc: SemanticDocument): Patch = {
+    lazy implicit val unit = global.newCompilationUnit(doc.input.text, doc.input.syntax)
+
+    val patchForExplicitResultTypes = addExplicitResultType()
+    val patchForTypeApply = AddTypeApply()
+
+    patchForExplicitResultTypes + patchForTypeApply
+  }
+
+  private def AddTypeApply()(implicit doc: SemanticDocument,
+                             unit: global.CompilationUnit): Patch = {
+    doc.synthetics.collect {
+      case syn@TypeApplyTree(function: SemanticTree, typeArguments: List[SemanticType]) =>
+        val symbols = typeArguments.map(getAbsoluteType)
+        if (symbols.length < typeArguments.length) Patch.empty
+        else {
+          val gsymbols = symbols.sequence
+          (for {
+            originalTree <- SyntheticHelper.getOriginalTree(syn)
+            if (!originalTree.isInstanceOf[Term.ApplyInfix])
+            if (!(originalTree.isInstanceOf[Pat] && !originalTree.isInstanceOf[Term])) // Never add types on elements that are on the right side of "="
+            replace <- if (syn.toString.startsWith("*.apply")) Some(".apply")
+            else if (syn.toString.startsWith("*[")) Some("")
+            else None
+            typesFromSemantic = gsymbols
+            typesSeenFromGlobal = getTypeNameAsSeenByGlobal(originalTree).map(_.map(_.dealiasWiden))
+            types <- typesSeenFromGlobal.orElse(typesFromSemantic)
+          } yield Patch.addRight(originalTree, s"${replace}[${types.mkString(", ")}]")).getOrElse(Patch.empty)
+        }
```

Comment:
In terms of performance it is better not to compute the `typesFromSemantic` too early because in most cases they are not needed.
Here is my suggestion:
```suggestion
               val patch = for {
          originalTree <- SyntheticHelper.getOriginalTree(syn)
          if (!originalTree.isInstanceOf[Term.ApplyInfix])
          if (!(originalTree.isInstanceOf[Pat] && !originalTree.isInstanceOf[Term])) // Never add types on elements that are on the right side of "="
          replace <- 
            if (syn.toString.startsWith("*.apply")) Some(".apply")
            else if (syn.toString.startsWith("*[")) Some("")
            else None
          typesSeenFromGlobal = getTypeNameAsSeenByGlobal(originalTree).map(_.map(_.dealiasWiden))
          types <- typesSeenFromGlobal.orElse(typeArguments.map(getAbsoluteType).sequence)
        } yield Patch.addRight(originalTree, s"${replace}[${types.mkString(", ")}]")
        patch.getOrElse(Patch.empty)
```

Reply:
yes I change it also